### PR TITLE
Make vocab update in get_docs deterministic

### PR DESCRIPTION
**Pull request by Stannislav**
The attribute `DocBin.strings` is a set. In `DocBin.get_docs`
a given vocab is updated by iterating over this set.
Iteration over a python set produces an arbitrary ordering,
therefore vocab is updated non-deterministically.

When training (fine-tuning) a spacy model, the base model's
vocabulary will be updated with the new vocabulary in the
training data in exactly the way described above. After
serialization, the file `model/vocab/strings.json` will
be sorted in an arbitrary way. This prevents reproducible
model training.

## Description

Here's a small code snippet that will reproduce the non-determinism.
```python
# spacy_strings.py
import sys

import spacy
from spacy.training import Corpus
from spacy.util import fix_random_seed


DATASET_PATH = "<some path>/train.spacy"


def run(prefix):
    # This shouldn't affect anything, just for the peace of mind
    fix_random_seed(7)

    nlp = spacy.blank("en")
    corpus = Corpus(DATASET_PATH)

    # This will update nlp's vocab as a side effect
    for _ in corpus(nlp):
        pass

    # Save the vocab to disk
    nlp.vocab.strings.to_disk(f"{prefix}_strings.json")


if __name__ == "__main__":
    run(sys.argv[1])

```
Run this this script twice and compare the results:
```bash
python spacy_strings.py 1
python spacy_strings.py 2
diff -q 1_strings.json 2_strings.json
# Files 1_strings.json and 2_strings.json differ
```
### Types of change
This is a bug fix.

## Checklist

- [x] I have submitted the spaCy Contributor Agreement.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Comment by svlandeg** [comment]
> After serialization, the file `model/vocab/strings.json` will
be sorted in an arbitrary way. This prevents reproducible
model training.

The file might not be serialized exactly the same way, but how does that influence the model training itself? In other words - are you actually seeing a difference in ML performance?

**Comment by Stannislav** [comment]
> Thanks for the PR! I see what you mean.
> 
> > After serialization, the file `model/vocab/strings.json` will
> > be sorted in an arbitrary way. This prevents reproducible
> > model training.
> 
> The file might not be serialized exactly the same way, but how does that influence the model training itself? In other words - are you actually seeing a difference in ML performance?

Ah, sorry, I didn't explain it properly.

I meant that the model checkpoint will have a different hash. So it's not obvious that the model is actually the same. This also breaks any hash-based tracking like DVC.

**Comment by svlandeg** [comment]
Ah, gotcha, thanks for clarifying!

I think the proposed edit looks fine. I guess it might cause a small delay when serializing, but it hopefully shouldn't be too bad? I do see the added advantage of ensuring model hash equality...

I'll leave this open a bit to allow my colleagues to chime in ;-)

**Comment by polm** [comment]
It looks good to me. I think having consistent output has a number of advantages.

Since vocabularies can be large I was concerned about extra memory use, but since the list of strings is a set I'm not sure that can be improved easily.

**Comment by honnibal** [comment]
I think we should make the `strings` variable a different data structure, probably a dict? I'm worried that the sorting can easily make a performance difference.
